Source/Core/DolphinQt2/MainWindow.cpp

```diff
@@ -478,3 +493,105 @@ void MainWindow::SetStateSlot(int slot)
   Settings::Instance().SetStateSlot(slot);
   m_state_slot = slot;
 }
+
+void MainWindow::PerformOnlineUpdate(const std::string& region)
+{
+  const int confirm = QMessageBox::question(
+      this, tr("Confirm"), tr("Connect to the Internet and perform an online system update?"));
+  if (confirm != QMessageBox::Yes)
+    return;
+
+  // Do not allow the user to close the dialog. Instead, wait until the update is finished
+  // or cancelled.
+  class UpdateProgressDialog final : public QProgressDialog
+  {
+  public:
+    using QProgressDialog::QProgressDialog;
+
+  protected:
+    void reject() override {}
+  };
+
+  UpdateProgressDialog dialog;
+  dialog.setLabelText(tr("Preparing to update...\nThis can take a while."));
+  dialog.setWindowTitle(tr("Updating"));
+  // QProgressDialog doesn't set its minimum size correctly.
+  dialog.setMinimumSize(360, 150);
+
+  // QProgressDialog doesn't allow us to disable the cancel button when it's pressed,
+  // so we have to pass it our own push button. Note: the dialog takes ownership of it.
+  auto* cancel_button = new QPushButton(tr("&Cancel"), this);
+  dialog.setCancelButton(cancel_button);
+  Common::Flag was_cancelled;
+  disconnect(&dialog, SIGNAL(canceled()), &dialog, SLOT(cancel()));
```

Comment:
The new, compile-time-safe way to connect/disconnect signals is like this:

```c++
disconnect(&dialog, &QProgressDialog::canceled, &dialog, &QProgressDialog::cancel);
```

Can you update it?